File: ctci/java/ch1-arrays-and-strings/q8_zero_matrix/Question.java
package q8_zero_matrix;

/**
 * Created by jjlee on 7/12/16.
 */
public class Question {
    /* in place using first row and first column to mark presence of zeros */
    public static void setZeros(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return;

        boolean firstRow = firstRowHasZeros(matrix);
        boolean firstColumn = firstColumnHasZeros(matrix);

        markZeros(matrix);

        for (int j = 1; j < matrix[0].length; j++) {
            if (matrix[0][j] == 0) {
                setColumnZeros(matrix, j);
            }
        }
        for (int i = 1; i < matrix.length; i++) {
            if (matrix[i][0] == 0) {
                setRowZeros(matrix, i);
            }
        }

        if (firstRow) setRowZeros(matrix, 0);
        if (firstColumn) setColumnZeros(matrix, 0);
    }

    /* check if first row has zeros */
    public static boolean firstRowHasZeros(int[][] matrix) {
        for (int j = 0; j < matrix[0].length; j++) {
            if (matrix[0][j] == 0) return true;
        }
        return false;
    }

    /* check if first column has zeros */
    public static boolean firstColumnHasZeros(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            if (matrix[i][0] == 0) return true;
        }
        return false;
    }

    /* mark zeros in first row and first column */
    public static void markZeros(int[][] matrix) {
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[i].length; j++) {
                if (matrix[i][j] == 0) {
                    matrix[0][j] = 0;
                    matrix[i][0] = 0;
                }
            }
        }
    }

    /* set given row to zeros */
    public static void setRowZeros(int[][] matrix, int row) {
        for (int j = 0; j < matrix[row].length; j++) {
            matrix[row][j] = 0;
        }
    }

    /* set given column to zeros */
    public static void setColumnZeros(int[][] matrix, int col) {
        for (int i = 0; i < matrix.length; i++) {
            matrix[i][col] = 0;
        }
    }
}

File: ctci/python/ch1-arrays-and-strings/q8-zero-matrix/Question.py
# in place using first row and first column to mark presence of zeros
def set_zeros(matrix):
    if len(matrix) == 0 or len(matrix[0]) == 0:
        return

    first_row = first_row_has_zeros(matrix)
    first_column = first_column_has_zeros(matrix)

    mark_zeros(matrix)

    for j in xrange(1, len(matrix[0])):
        if matrix[0][j] == 0:
            set_column_zeros(matrix, j)
    for i in xrange(1, len(matrix)):
        if matrix[i][0] == 0:
            set_row_zeros(matrix, i)

    if first_row:
        set_row_zeros(matrix, 0)
    if first_column:
        set_column_zeros(matrix, 0)


# check if first row has zeros
def first_row_has_zeros(matrix):
    for j in xrange(len(matrix[0])):
        if matrix[0][j] == 0:
            return True
    return False


# check if first column has zeros
def first_column_has_zeros(matrix):
    for i in xrange(len(matrix)):
        if matrix[i][0] == 0:
            return True
    return False


# mark zeros in first row and first column
def mark_zeros(matrix):
    for i in xrange(1, len(matrix)):
        for j in xrange(1, len(matrix[i])):
            if matrix[i][j] == 0:
                matrix[0][j] = 0
                matrix[i][0] = 0


# set given row to zeros
def set_row_zeros(matrix, row):
    for j in xrange(len(matrix[0])):
        matrix[row][j] = 0


# set given column to zeros
def set_column_zeros(matrix, column):
    for i in xrange(len(matrix)):
        matrix[i][column] = 0
